### PR TITLE
switch to front-coded v1 bucket size 4 by default

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
I think we should consider switching the `IndexSpec` default value of `stringDictionaryEncoding` to `{"type":"frontCoded", "bucketSize":4, "formatVersion":1}`.

Based on measurements #13854 things look pretty good and we have been running version 0 of the format for some time on a number of datasources without any notable performance loss, and version 1 for a smaller amount of time. I think by the time 27 is released it should be sufficiently baked in to feel confident about it being the default.

However, this means that upgrading from versions older than 26 will need special consideration, so it is important to call out in the release notes if we go forward with this.

#### Release note
Front coding was originally introduced in Druid 25.0, and an improved 'version 1' was introduced in Druid 26.0, with typically faster read speed and smaller storage size, has become the default in Druid 27.0. This means by default, segments created with Druid 27.0 are backwards compatible with Druid 26.0, but not compatible with Druid versions older than 26.0. If upgrading to Druid 27.0 from a version older than 26.0, the `stringDictionaryEncoding` should be set to `{"type": "utf8"}` to keep writing out the older format to enable seamless downgrades to Druid 25.0 and older, and then later is recommended to be changed to the new default once determined that rollback is not necessary.

<hr>

This PR has:

- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by gianm** [comment]
People may find it difficult to update all their ingest specs, supervisors, etc, to ensure backwards compatibility. Is it possible to also add a runtime property that controls the default? That way, a cluster admin only has to set it in one place rather than track down all their users.

File: docs/ingestion/ingestion-spec.md [hunk 1]
```diff
@@ -495,18 +495,18 @@ The `indexSpec` object can include the following properties:
...
+|stringDictionaryEncoding|Encoding format for STRING value dictionaries used by STRING and COMPLEX&lt;json&gt; columns. <br /><br />Example to enable front coding: `{"type":"frontCoded", "bucketSize": 4}`<br />`bucketSize` is the number of values to place in a bucket to perform delta encoding. Must be a power of 2, maximum is 128. Defaults to 4.<br /> `formatVersion` can specify older versions for backwards compatibility during rolling upgrades, valid options are `0` and `1`, defaults to `1`.<br /><br />See [Front coding](#front-coding) for more information.|`{"type":"frontCoded", "bucketSize": 4, "formatVersion": 1}`|
```
Comment (ektravel) [inline docs/ingestion/ingestion-spec.md:498]:
```suggestion
|stringDictionaryEncoding|Encoding format for STRING value dictionaries used by STRING and COMPLEX&lt;json&gt; columns. <br /><br />Example to enable front coding: `{"type":"frontCoded", "bucketSize": 4}`<br />`bucketSize` is the number of values to place in a bucket to perform delta encoding. Must be a power of 2, maximum is 128. Defaults to 4.<br /> `formatVersion` can specify older versions for backwards compatibility during rolling upgrades. Valid options are 0 and 1. Defaults to 1.<br /><br />See [Front coding](#front-coding) for more information.|`{"type":"frontCoded", "bucketSize": 4, "formatVersion": 1}`|
```

File: docs/ingestion/ingestion-spec.md [hunk 2]
```diff
@@ -495,18 +495,18 @@ The `indexSpec` object can include the following properties:
...
+> Front coding was originally introduced in Druid 25.0, and an improved 'version 1' was introduced in Druid 26.0, with typically faster read speed and smaller storage size, before finally becoming the default in Druid 27.0. By default, segments created with Druid 27.0 are backwards compatible with Druid 26.0, but not compatible with Druid versions older than 26.0. If upgrading to Druid 27.0 from a version older than 26.0, the `stringDictionaryEncoding` should be set to `{"type": "utf8"}` to keep writing out the older format to enable seamless downgrades to Druid 25.0 and older, and then later is recommended to be changed to the new default once determined that rollback is not necessary.
```
Comment (ektravel) [inline docs/ingestion/ingestion-spec.md:509]:
```suggestion
> Front coding was originally introduced in Druid 25.0. Then, an improved 'version 1' was introduced in Druid 26.0, with typically faster read speed and smaller storage size, before finally becoming the default in Druid 27.0. By default, segments created with Druid 27.0 are backwards compatible with Druid 26.0, but not compatible with Druid versions older than 26.0. If upgrading to Druid 27.0 from a version older than 26.0, set the `stringDictionaryEncoding` to `{"type": "utf8"}` to keep writing out the older format to enable seamless downgrades to Druid 25.0 and older, and then later is recommended to be changed to the new default once determined that rollback is not necessary.
```
This part is a bit confusing:
"...and then later is recommended to be changed to the new default once determined that rollback is not necessary."
Are we recommending that users set `stringDictionaryEncoding` to front coding?

**Review by ektravel — COMMENTED** [review]
Left a few suggestions related to the documentation.

**Pull request closed without merging**